#!/usr/bin/python

# validannot_env.py
# Define paths to all the directories needed
# Needs to be included in all the validannot scripts
root_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/"
gff3_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/gff3/"
dna_fasta_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/dna_fasta/"
cdna_fasta_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/cdna_fasta/"
ncrna_fasta_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/ncrna_fasta/"
gtf_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/gtf/"
log_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/log/"
biomart_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/biomart/"
validation_path = "$HOME/shares-net/sequencages/ressources/validation_genomeannot/validation/"



